### PR TITLE
Add a Prometheus Remote Write connector to the exporting engine

### LINKED ISSUE #7175: Implement Prometheus Remote Write exporting connector
Feature idea summary
### Feature summary

* Code already exists in the backend that can export Prometheus.
* Testing client already exists.
* Code currently has poor structure - intertwined with the existing backends.
* Sits on top of two libraries in C++ (protobuf and snappy), unknown if this wrapper will change.
* Binary protocol rather than text.
* Unit-tests will need to call the wrapped C++ code - unclear if this is difficult.

### Expected behavior

* Reviewer will have a testing client.
* Note for reviewer - the threading/locking should be standard for the exporting engine, may need to read the config reading code.

### PR DESCRIPTION
**Pull request by vlvkobal**
##### Summary
Implement a Prometheus Remote Write exporting connector as a part of the backends refactoring project #6762.

No need to check code in the first commit "Copy files for the Prometheus remote write exporting connector". I copied the files from the backends.

Closes #7175

##### Component Name
backends, exporting engine

##### Description of testing that the developer performed
Tested with [Remote Write adapter example](https://github.com/prometheus/prometheus/tree/master/documentation/examples/remote_storage/example_write_adapter)
`exporting.conf`:
```
[prometheus_remote_write:instance6]
enabled = yes
destination = localhost:1234
remote write URL path = /receive
data source = as collected
update every = 2
send charts matching = system.processes

[prometheus_remote_write:instance7]
enabled = yes
destination = localhost:1235
remote write URL path = /receive
data source = average
update every = 2
send charts matching = system.cpu
```

### CODE REVIEW

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- backends/prometheus/remote_write/remote_write.cc  4
- backends/prometheus/backend_prometheus.c  21
- backends/backends.c  2
- exporting/prometheus/remote_write/remote_write.c  5
- exporting/prometheus/remote_write/remote_write_request.cc  6
- exporting/prometheus/prometheus.c  29
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5121710&bid=16837821)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- backends/backends.c  2
- exporting/prometheus/remote_write/remote_write.c  6
- exporting/prometheus/prometheus.c  28
- backends/prometheus/backend_prometheus.c  21
- exporting/prometheus/remote_write/remote_write_request.cc  6
- backends/prometheus/remote_write/remote_write.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5121710&bid=16837821)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- backends/prometheus/remote_write/remote_write.cc  4
- exporting/prometheus/prometheus.c  28
- exporting/prometheus/remote_write/remote_write_request.cc  6
- backends/prometheus/backend_prometheus.c  21
- backends/backends.c  2
- exporting/prometheus/remote_write/remote_write.c  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5121710&bid=16837821)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- exporting/aws_kinesis/aws_kinesis.c  1
- exporting/prometheus/remote_write/remote_write_request.cc  6
- exporting/prometheus/remote_write/remote_write.c  6
- exporting/prometheus/prometheus.c  28
- backends/backends.c  2
- backends/prometheus/backend_prometheus.c  21
- backends/prometheus/remote_write/remote_write.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5121710&bid=16837821)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- backends/prometheus/remote_write/remote_write.cc  4
- backends/backends.c  2
- exporting/prometheus/prometheus.c  28
- backends/prometheus/backend_prometheus.c  21
- exporting/prometheus/remote_write/remote_write.c  6
- exporting/aws_kinesis/aws_kinesis.c  1
- exporting/prometheus/remote_write/remote_write_request.cc  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5121710&bid=16837821)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- exporting/prometheus/remote_write/remote_write_request.cc  6
- exporting/aws_kinesis/aws_kinesis.c  1
- backends/prometheus/backend_prometheus.c  20
- exporting/prometheus/remote_write/remote_write.c  5
- backends/prometheus/remote_write/remote_write.cc  4
- exporting/json/json.c  1
- exporting/opentsdb/opentsdb.c  2
- backends/backends.c  2
- exporting/prometheus/prometheus.c  28
- exporting/graphite/graphite.c  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5121710&bid=16837821)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- exporting/prometheus/prometheus.c  28
- exporting/aws_kinesis/aws_kinesis.c  1
- backends/backends.c  2
- exporting/prometheus/remote_write/remote_write_request.cc  6
- exporting/graphite/graphite.c  1
- exporting/opentsdb/opentsdb.c  2
- exporting/prometheus/remote_write/remote_write.c  4
- backends/prometheus/remote_write/remote_write.cc  4
- exporting/json/json.c  1
- backends/prometheus/backend_prometheus.c  19
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5121710&bid=16837821)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- exporting/json/json.c  1
- exporting/prometheus/prometheus.c  28
- exporting/opentsdb/opentsdb.c  2
- backends/prometheus/backend_prometheus.c  19
- exporting/graphite/graphite.c  1
- exporting/prometheus/remote_write/remote_write_request.cc  6
- backends/backends.c  2
- backends/prometheus/remote_write/remote_write.cc  4
- exporting/aws_kinesis/aws_kinesis.c  1
- exporting/prometheus/remote_write/remote_write.c  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5121710&bid=16837821)

File: exporting/init_connectors.c [hunk 1]
```diff
@@ -37,6 +38,12 @@ int init_connectors(struct engine *engine)
                 if (init_opentsdb_http_instance(instance) != 0)
                     return 1;
                 break;
+            case BACKEND_TYPE_PROMETHEUS_REMOTE_WRITE:
+#if ENABLE_PROMETHEUS_REMOTE_WRITE
```
Comment (thiagoftsm) [inline exporting/init_connectors.c]: Considering that when ` ENABLE_PROMETHEUS_REMOTE_WRITE` is not set we won't return 1 never, should you not put the `case` and `break` inside the `#if .. #endif`.

File: exporting/prometheus/remote_write/README.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
+# Prometheus remote write exporting connector
+
+## Prerequisites
+
+To use the Prometheus remote write API with [storage
+providers](https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage) you should install first
```
Comment (shortpatti) [inline exporting/prometheus/remote_write/README.md]:
```suggestion
providers](https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage), install
```

File: exporting/prometheus/remote_write/README.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+[protobuf](https://developers.google.com/protocol-buffers/) and [snappy](https://github.com/google/snappy) libraries.
+Next, Netdata should be re-installed from the source. The installer will detect that the required libraries and
```
Comment (shortpatti) [inline exporting/prometheus/remote_write/README.md]:
```suggestion
Next, re-install Netdata from the source, which detects that the required libraries and
```

File: exporting/init_connectors.c (hunk 1 above)
Comment (vlvkobal) [reply exporting/init_connectors.c]: I don't want it neither printing "EXPORTING: unknown exporting connector type" nor returning 1 in that case. The "Prometheus Remote Write support isn't compiled" message was already printed before the connectors initialization started.

File: exporting/prometheus/README.md [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+
+```conf
+[backend]
+	send charts matching = *
+```
+
+This settings accepts a space separated list of patterns to match the **charts** to be sent to Prometheus. Each pattern
+can use `*` as wildcard, any number of times (e.g `*a*b*c*` is valid). Patterns starting with `!` give a negative match
+(e.g `!*.bad users.* groups.*` will send all the users and groups except `bad` user and `bad` group). The order is
+important: the first match (positive or negative) left to right, is used.
```
Comment (thiagoftsm) [inline exporting/prometheus/README.md]: This is a simple pattern according our code: 

```
charts_pattern = simple_pattern_create(config_get(CONFIG_SECTION_BACKEND, "send charts matching", "*"), NULL, SIMPLE_PATTERN_EXACT);
```

please, link this paragraph to the simple pattern doc.

Comment (vlvkobal) [reply exporting/prometheus/README.md]: Done

File: exporting/init_connectors.c (hunk 1 above)
Comment (thiagoftsm) [reply exporting/init_connectors.c]: I got it, thanks!

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- exporting/graphite/graphite.c  1
- exporting/prometheus/prometheus.c  28
- exporting/aws_kinesis/aws_kinesis.c  1
- exporting/prometheus/remote_write/remote_write_request.cc  6
- backends/backends.c  2
- exporting/prometheus/remote_write/remote_write.c  4
- exporting/opentsdb/opentsdb.c  2
- backends/prometheus/remote_write/remote_write.cc  4
- exporting/json/json.c  1
- backends/prometheus/backend_prometheus.c  19
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5121710&bid=16837821)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- exporting/aws_kinesis/aws_kinesis.c  1
- exporting/json/json.c  1
- exporting/prometheus/remote_write/remote_write_request.cc  6
- exporting/prometheus/prometheus.c  28
- exporting/opentsdb/opentsdb.c  2
- backends/prometheus/remote_write/remote_write.cc  4
- exporting/tests/test_exporting_engine.c  2
- exporting/prometheus/remote_write/remote_write.c  4
- backends/prometheus/backend_prometheus.c  19
- exporting/graphite/graphite.c  1
- backends/backends.c  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5121710&bid=16837821)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- backends/prometheus/remote_write/remote_write.cc  4
- backends/prometheus/backend_prometheus.c  19
- exporting/prometheus/remote_write/remote_write.c  4
- exporting/aws_kinesis/aws_kinesis.c  1
- backends/backends.c  2
- exporting/json/json.c  1
- exporting/prometheus/remote_write/remote_write_request.cc  6
- exporting/tests/test_exporting_engine.c  2
- exporting/opentsdb/opentsdb.c  2
- exporting/graphite/graphite.c  1
- exporting/prometheus/prometheus.c  28
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5121710&bid=16837821)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- exporting/tests/test_exporting_engine.c  2
- backends/prometheus/backend_prometheus.c  19
- exporting/prometheus/remote_write/remote_write_request.cc  6
- exporting/opentsdb/opentsdb.c  2
- exporting/prometheus/prometheus.c  28
- exporting/graphite/graphite.c  1
- exporting/prometheus/remote_write/remote_write.c  4
- backends/backends.c  2
- backends/prometheus/remote_write/remote_write.cc  4
- exporting/aws_kinesis/aws_kinesis.c  1
- exporting/json/json.c  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5121710&bid=16837821)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- exporting/json/json.c  1
- exporting/aws_kinesis/aws_kinesis.c  1
- exporting/opentsdb/opentsdb.c  2
- backends/prometheus/remote_write/remote_write.cc  4
- exporting/prometheus/remote_write/remote_write.c  4
- backends/backends.c  2
- backends/prometheus/backend_prometheus.c  19
- exporting/prometheus/remote_write/remote_write_request.cc  6
- exporting/prometheus/prometheus.c  28
- exporting/tests/test_exporting_engine.c  2
- exporting/graphite/graphite.c  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5121710&bid=16837821)

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- exporting/tests/test_exporting_engine.c  2
- exporting/prometheus/remote_write/remote_write_request.cc  6
- backends/backends.c  2
- exporting/prometheus/remote_write/remote_write.c  4
- exporting/aws_kinesis/aws_kinesis.c  1
- exporting/opentsdb/opentsdb.c  2
- exporting/json/json.c  1
- exporting/graphite/graphite.c  1
- backends/prometheus/backend_prometheus.c  19
- exporting/prometheus/prometheus.c  27
- backends/prometheus/remote_write/remote_write.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5121710&bid=16837821)

File: exporting/prometheus/remote_write/README.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+To use the Prometheus remote write API with [storage
+providers](https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage), install
+[protobuf](https://developers.google.com/protocol-buffers/) and [snappy](https://github.com/google/snappy) libraries.
+Next, re-install Netdata from the source, which detects that the required libraries and
+utilities are now available.
+
+## Configuration
+
+To enable data sending to a storage provider using the Prometheus remote write API, run `./edit-config exporting.conf`
```
Comment (joelhans) [inline exporting/prometheus/remote_write/README.md]: We should stick with using the word "exporting" when talking about these connectors.

```suggestion
To enable data exporting to a storage provider using the Prometheus remote write API, run `./edit-config exporting.conf`
```

File: exporting/prometheus/remote_write/README.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
+# Prometheus remote write exporting connector
+
+## Prerequisites
+
+To use the Prometheus remote write API with [storage
+providers](https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage), install
+[protobuf](https://developers.google.com/protocol-buffers/) and [snappy](https://github.com/google/snappy) libraries.
+Next, re-install Netdata from the source, which detects that the required libraries and
```
Comment (joelhans) [inline exporting/prometheus/remote_write/README.md]: Could you also re-install using a kickstart script, or does it have to be from source?

File: exporting/prometheus/remote_write/README.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+```conf
+[remote_write:my_instance]
+    enabled = yes
+    destination = example.domain:example_port
+    remote write URL path = /receive
+```
+
+`remote write URL path` is used to set an endpoint path for the remote write protocol. The default value is `/receive`.
+For example, if your endpoint is `http://example.domain:example_port/storage/read` you should set
```
Comment (joelhans) [inline exporting/prometheus/remote_write/README.md]:
```suggestion
For example, if your endpoint is `http://example.domain:example_port/storage/read`:
```

File: exporting/prometheus/remote_write/README.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
+# Prometheus remote write exporting connector
+
```
Comment (joelhans) [inline exporting/prometheus/remote_write/README.md]: Could we add a new sentence here that very briefly explains what this connector does?

The Prometheus remote write exporting connector uses the [exporting engine](../README.md) to send Netdata metrics to your choice of more than 20 external storage providers for long-term archiving and further analysis.

File: exporting/prometheus/remote_write/README.md (hunk 6 above)
Comment (vlvkobal) [reply exporting/prometheus/remote_write/README.md]: AFAIK `kickstart.sh` doesn't install the dependencies needed for the remote write connector.

File: exporting/prometheus/remote_write/README.md (hunk 8 above)
Comment (vlvkobal) [reply exporting/prometheus/remote_write/README.md]: Done

**Comment by joelhans** [comment]
I didn't realize until just now that you have the big `exporting/prometheus/README.md` file that you've duplicated in this PR. What's the value of having that document in two places?

**Comment by vlvkobal** [comment]
I'm preparing the Prometheus Web API for transferring to the exporting engine code. It is not used right now by the Web API, but the remote write connector uses its code.

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- exporting/graphite/graphite.c  1
- exporting/tests/test_exporting_engine.c  2
- backends/prometheus/remote_write/remote_write.cc  4
- exporting/json/json.c  1
- backends/prometheus/backend_prometheus.c  19
- exporting/prometheus/remote_write/remote_write.c  4
- exporting/prometheus/remote_write/remote_write_request.cc  6
- exporting/aws_kinesis/aws_kinesis.c  1
- exporting/prometheus/prometheus.c  27
- exporting/opentsdb/opentsdb.c  2
- backends/backends.c  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5121710&bid=16837821)

File: exporting/prometheus/remote_write/remote_write.c [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+        "POST %s HTTP/1.1\r\n"
+        "Host: %s\r\n"
+        "Accept: */*\r\n"
+        "Content-Length: %zu\r\n"
+        "Content-Type: application/x-www-form-urlencoded\r\n\r\n",
+        connector_specific_config->remote_write_path,
+        instance->engine->config.hostname,
+        buffer_strlen((BUFFER *)instance->buffer));
+
+    send(*sock, buffer_tostring(header), buffer_strlen(header), flags);
```
Comment (stelfrag) [inline exporting/prometheus/remote_write/remote_write.c]: You are not checking the return size of the send to make sure the bytes are written

Comment (vlvkobal) [reply exporting/prometheus/remote_write/remote_write.c]: Fixed.

**Comment by ktsaou** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- exporting/prometheus/remote_write/remote_write_request.cc  6
- backends/prometheus/remote_write/remote_write.cc  4
- backends/prometheus/backend_prometheus.c  19
- exporting/prometheus/prometheus.c  27
- exporting/graphite/graphite.c  1
- exporting/opentsdb/opentsdb.c  2
- exporting/prometheus/remote_write/remote_write.c  4
- exporting/json/json.c  1
- backends/backends.c  2
- exporting/aws_kinesis/aws_kinesis.c  1
- exporting/tests/test_exporting_engine.c  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/netdata/netdata/pullRequest?prid=5121710&bid=16837821)

**Pull request merged**